### PR TITLE
feat(backend/copilot): async polling for agent-generator + SSE auto-reconnect

### PR DESCRIPTION
**Pull request by majdyz**
## Summary
- Platform service client now uses submit-and-poll pattern for all agent-generator calls (10s poll interval, 30min timeout via `time.monotonic()`)
- `asyncio.sleep()` in poll loop yields to event loop, keeping SSE heartbeats alive through GCP's L7 load balancer
- Transient HTTP errors (429, 503, 504, 408) and connection errors are retried (up to 5 consecutive), not fatal
- `agent_json` responses validated before returning (prevents silent `None` propagation)
- Frontend auto-reconnects up to 3 times when SSE drops mid-stream
- `agentgenerator_timeout` setting wired into HTTP client (was hardcoded)

## Why
GCP's L7 load balancer kills SSE connections at ~5 minutes. Agent generation takes 5–30 minutes, so the old synchronous HTTP call would block for the entire duration, causing the SSE connection to die. The new polling pattern keeps each request short (~30s timeout) while `asyncio.sleep(10)` yields control for heartbeat delivery.

## Changes
| File | Change |
|---|---|
| `service.py` | `_submit_and_poll()` helper with retry logic for transient errors, `agent_json` validation, `agentgenerator_timeout` wired into client, `time.monotonic()` for wall-clock timeout |
| `settings.py` | `agentgenerator_timeout` default 1800→30 (now per-request, not total) |
| `useCopilotPage.ts` | Auto-reconnect on mid-stream SSE drop (max 3 attempts, 1s delay) |
| `test_service.py` | 35 tests: polling happy path, failures, retry, timeout, transient HTTP retry, missing agent_json validation |

## Companion PR
- Agent-generator async endpoints + Redis job store: https://github.com/Significant-Gravitas/AutoGPT-Agent-Generator/pull/120

## Review comments addressed
- [x] **autogpt-reviewer**: Wire `agentgenerator_timeout` setting into HTTP client
- [x] **autogpt-reviewer**: Use `time.monotonic()` for accurate wall-clock timeout
- [x] **autogpt-reviewer**: Restore `decompose_goal_external` docstring with return-type enumeration
- [x] **coderabbitai**: Retry transient HTTP statuses (429/503/504/408) during polling
- [x] **coderabbitai**: Validate `agent_json` exists before returning from 3 endpoints

## Test plan
- [x] 35/35 backend tests pass
- [x] CI green (23/23 checks)
- [ ] Deploy both PRs to dev and test full agent generation flow
- [ ] Test SSE drop recovery: wait for LB kill at 5min → verify frontend reconnects
- [ ] Verify polling heartbeats keep SSE alive during long generation

### CODE REVIEW

**Review by autogpt-reviewer — COMMENTED** [review]
**PR #12199 — feat(copilot): async polling for agent-generator + SSE auto-reconnect**
Author: majdyz (Zamil Majdy) | Files: `service.py` (+183/-239), `settings.py` (+2/-2), `useCopilotPage.ts` (+40/-1), `test_service.py` (+372/-203)

**🎯 Verdict: APPROVE WITH CONDITIONS**

---

### What This PR Does
GCP's L7 load balancer kills SSE connections at ~5 minutes, but agent generation takes 5–30 minutes. This PR switches the platform backend from a single long-lived HTTP call to a submit-and-poll pattern (POST → 202 + job_id → GET poll every 10s, 30min budget). The frontend adds auto-reconnect (3 attempts, 1s delay) when SSE drops mid-stream. Together, these two layers solve the connection-drop problem at both backend and frontend.

### Specialist Findings

🛡️ **Security** ✅ — No blockers. `job_id` flows entirely within backend-to-microservice boundary — no IDOR risk. SSE reconnect reuses existing authenticated session — no auth bypass. Minor nit: `job_id` is interpolated directly into URL path (`service.py:186`) without format validation; a UUID/alphanumeric check would harden against hypothetical path traversal from a compromised internal service.

🏗️ **Architecture** ✅ — Clean, well-isolated abstraction. `_submit_and_poll()` has single responsibility; all 4 external functions delegate uniformly. Massive DRY win — eliminated 4× identical try/except blocks. Frontend and backend layers are independent (good separation). ⚠️ `agentgenerator_timeout` setting (`settings.py:374`) is now dead code — `_REQUEST_TIMEOUT` is hardcoded at `service.py:82`. Wire it back or delete the setting.

⚡ **Performance** ✅ — `asyncio.sleep(10)` correctly yields to event loop for heartbeat delivery. Singleton `httpx.AsyncClient` with connection pooling is efficient. No memory leaks in poll loop. ⚠️ Fixed 10s interval = up to 180 requests per generation; exponential backoff (2s→30s cap) would reduce to ~50 while detecting fast jobs sooner. ⚠️ Elapsed time tracking (`service.py:190`) drifts from wall clock — uses counter instead of `time.monotonic()`.

🧪 **Testing** ⚠️ — Backend `TestSubmitAndPoll` (8 tests) is solid: happy path, failures, transient retry, timeout all covered. Public function tests properly mock at the `_submit_and_poll` boundary. **However:** ❌ Frontend SSE auto-reconnect (40 lines of stateful `useEffect` logic) has **zero test coverage** — this is the user-facing reliability fix. ❌ `customize_template_external` has no test class at all (missing `TestCustomizeTemplateExternal`). Missing edge cases: all-transient-errors-until-timeout, malformed poll JSON, empty `result` on completed job.

📖 **Quality** ✅ — Genuinely cleaner refactor (net -56 lines). Good naming conventions (`_submit_and_poll`, `POLL_INTERVAL_SECONDS`). Well-organized section comments. ⚠️ Docstrings trimmed too aggressively — `decompose_goal_external` lost useful return-type enumeration (`instructions`/`clarifying_questions`/`unachievable_goal`/`vague_goal`/`error`). ⚠️ `setTimeout(() => resumeStream(), 1_000)` — extract as `RECONNECT_DELAY_MS` for consistency.

📦 **Product** ✅ — Solves a real, painful user problem. Happy path is completely transparent to users. Reconnect is well-guarded with attempt limits and proper cleanup. Toast on exhaustion is clear. ⚠️ No progress indicator during long generation (25 min with zero feedback). ⚠️ Unclear if Stop button cancels the `_submit_and_poll` loop. ⚠️ Consider "Reconnecting..." indicator instead of brief error-state flash during 1s reconnect delay.

📬 **Discussion** ⚠️ — Zero human reviews (PR opened ~3 hours ago). Both Greptile and CodeRabbit skipped due to inflated diff (299 files from dev merge — only 4 actual changes). 🔴 Merge conflict with #12071 (stale/approved). Companion PR dependency: AutoGPT-Agent-Generator/pull/120 must deploy alongside for polling endpoints to exist. All 5 manual test plan checkboxes unchecked.

🔎 **QA** ⚠️ — Environment partially broken (backend/Supabase down, only Next.js frontend running). Login blocked, copilot page inaccessible. Code inspection confirms reconnect logic is well-implemented. Could not perform live SSE reconnect testing.

---

### Conditions for Merge

1. **`agentgenerator_timeout` setting is dead code** — `settings.py:374-376` defines it, `service.py:82` hardcodes `_REQUEST_TIMEOUT = httpx.Timeout(30.0)` instead. Either wire `settings.config.agentgenerator_timeout` into the timeout, or delete the setting. Dead config misleads operators. *(Flagged by: Security, Architect, Quality)*

2. **Restore `decompose_goal_external` return-type documentation** — The docstring lost the enumeration of possible return types (`instructions`, `clarifying_questions`, `unachievable_goal`, `vague_goal`, `error`). One-line abbreviated `Args:` sections for all 4 functions would also help callers. *(Flagged by: Quality)*

### Should Fix (Follow-up OK)

1. **Frontend SSE reconnect needs tests** — 40 lines of stateful `useEffect` logic with `reconnectAttemptsRef`, `setTimeout` cleanup, counter reset. At minimum: RTL test verifying reconnect triggers on `streaming→error` + `hasActiveStream`, counter exhaustion shows toast, counter resets on success. *(Flagged by: Testing)*
2. **Add `TestCustomizeTemplateExternal`** — Missing test class entirely. Function has its own response-mapping logic. *(Flagged by: Testing)*
3. **Exponential backoff for polling** — Fixed 10s = 180 requests worst case. Start at 2s, cap at 30s → ~50 requests, faster detection of quick jobs. *(Flagged by: Performance, Architect)*
4. **Use `time.monotonic()` for elapsed tracking** — Current counter drifts by HTTP request duration per iteration. *(Flagged by: Performance, Quality)*
5. **Extract `RECONNECT_DELAY_MS = 1_000`** — Inline magic number in frontend. *(Flagged by: Quality)*
6. **Rebase to reduce diff noise** — 299-file diff from dev merge blocks automated reviewers. 4 actual files. *(Flagged by: Discussion)*

### Risk Assessment
Merge risk: **LOW** | Rollback: **EASY** (revert 4 files, no migrations)

**Deployment note:** Requires companion PR (AutoGPT-Agent-Generator/pull/120) deployed alongside for polling endpoints. Without it, backend will get unexpected responses from agent-generator. Dummy mode still works independently.

@ntindle Two small conditions (dead setting + restore docstring), then this is good to go. Backend refactor is a clean DRY win, frontend reconnect is well-implemented. Main gap is frontend test coverage — not blocking merge but should be a fast follow-up.

**Comment by majdyz** [comment]
## Review Items Addressed (0bc098acb)

**Must Fix (both done):**
1. ✅ Wired `agentgenerator_timeout` setting into HTTP client — `_get_client()` now reads `settings.config.agentgenerator_timeout` dynamically instead of the hardcoded constant
2. ✅ Replaced `elapsed += POLL_INTERVAL_SECONDS` with `time.monotonic()` delta for accurate wall-clock timeout tracking

**Should Fix (done in same commit):**
3. ✅ Added `MAX_CONSECUTIVE_POLL_ERRORS = 5` — polling loop now gives up after 5 consecutive transient `RequestError` failures instead of retrying for the full 30-min budget
4. ✅ Restored `decompose_goal_external` docstring with full return-type enumeration

**Tests updated:** `test_poll_timeout` now mocks `time.monotonic`, new `test_poll_gives_up_after_consecutive_transient_errors` test added. All 32 tests pass.

File: autogpt_platform/backend/backend/copilot/tools/agent_generator/service.py [hunk 1]
```diff
@@ -136,39 +127,127 @@ def _get_client() -> httpx.AsyncClient:
...
+            poll_resp.raise_for_status()
+        except httpx.HTTPStatusError as e:
+            if e.response.status_code == 404:
+                return _create_error_response(
+                    "Agent Generator job not found or expired", "job_not_found"
+                )
+            error_type, error_msg = _classify_http_error(e)
+            logger.error(f"Poll error for job {job_id}: {error_msg}")
+            return _create_error_response(error_msg, error_type)
+        except httpx.RequestError as e:
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/tools/agent_generator/service.py:212]: Fixed in 9b3e25d — transient HTTP statuses (429, 503, 504, 408) now increment `consecutive_errors` and retry, same as `RequestError`. Added `test_poll_retries_on_transient_http_status` and `test_poll_does_not_retry_non_transient_http_status` tests.

File: autogpt_platform/backend/backend/copilot/tools/agent_generator/service.py [hunk 2]
```diff
@@ -177,119 +256,73 @@ async def decompose_goal_external(
...
         error_msg = f"Unexpected error calling Agent Generator: {e}"
         logger.error(error_msg)
         return _create_error_response(error_msg, "unexpected_error")
 
+    if result.get("type") == "error":
+        return result
+
+    # The job result contains {"agent_json": {...}, "success": True, ...}
+    return result.get("agent_json")
+
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/tools/agent_generator/service.py:364]: Fixed in 9b3e25d — all 3 `result.get("agent_json")` sites now validate the value is a dict, returning `_create_error_response(..., "invalid_response")` otherwise. Added `test_generate_agent_missing_agent_json` test.

File: autogpt_platform/backend/backend/copilot/tools/agent_generator/service.py [hunk 3]
```diff
@@ -136,39 +127,142 @@ def _get_client() -> httpx.AsyncClient:
...
+                return _create_error_response(error_msg, "poll_error")
+            continue
+
+        consecutive_errors = 0
+        poll_data = poll_resp.json()
+        status = poll_data.get("status")
+
+        if status == "completed":
+            logger.info(f"Agent Generator job {job_id} completed")
+            return poll_data.get("result", {})
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/tools/agent_generator/service.py]: Fixed in 8ef8bec — `_submit_and_poll` now validates `poll_data["result"]` is a dict before returning, returns `invalid_response` error otherwise.

**Comment by majdyz** [comment]
Addressed both nitpicks from the latest coderabbit review in ec7c7eb:

1. **Code dedup** — extracted `_extract_agent_json(result)` helper, used in all 3 agent-producing endpoints.
2. **Unknown poll status** — now explicitly handles `running`/`pending`/`queued` as known running states; unknown statuses fail fast with `invalid_response` instead of polling for 30 minutes.

**Comment by ntindle** [comment]
is this still needed?
